File: services/control-plane/app/workspace_agents/task_dependencies.py
"""Shared task dependency checks for lease, dispatch, and operator start."""

from __future__ import annotations

from typing import Any

from app.persistence import task_store


def dependencies_completed(task: dict[str, Any] | None) -> bool:
    if not isinstance(task, dict):
        return True
    return not dependency_blockers(task)


def dependency_blockers(task: dict[str, Any] | None) -> list[dict[str, str]]:
    if not isinstance(task, dict):
        return []
    deps = task.get("dependencies")
    if not isinstance(deps, list) or not deps:
        return []
    blockers: list[dict[str, str]] = []
    for dep_id in deps:
        cleaned = str(dep_id or "").strip()
        if not cleaned:
            continue
        dep = task_store.get_task(cleaned)
        if dep is None:
            blockers.append(
                {
                    "task_id": cleaned,
                    "status": "missing",
                    "owner_role": "",
                    "summary": "missing dependency",
                }
            )
            continue
        status = str(dep.get("status") or "").strip().lower()
        if status != "completed":
            goal = str(dep.get("goal") or "").strip()
            blockers.append(
                {
                    "task_id": cleaned,
                    "status": status or "unknown",
                    "owner_role": str(dep.get("owner_role") or "").strip().lower(),
                    "summary": goal[:120] if goal else cleaned,
                }
            )
    return blockers


def dependency_blocker_message(task: dict[str, Any] | None) -> str:
    blockers = dependency_blockers(task)
    if not blockers:
        return ""
    parts: list[str] = []
    for row in blockers[:3]:
        role = row.get("owner_role") or "task"
        status = row.get("status") or "unknown"
        task_id = row.get("task_id") or ""
        parts.append(f"{role} {task_id} ({status})")
    suffix = f" (+{len(blockers) - 3} more)" if len(blockers) > 3 else ""
    return f"blocked by unfinished dependencies: {', '.join(parts)}{suffix}"


__all__ = [
    "dependencies_completed",
    "dependency_blocker_message",
    "dependency_blockers",
]

File: apps/console-web/src/lib/task-dependencies.ts
import type { WorkspaceTaskRecord } from '../api/tasks-api';

export type TaskDependencyBlocker = {
  taskId: string;
  status: string;
  ownerRole: string;
  summary: string;
};

export function taskDependencyBlockers(
  task: WorkspaceTaskRecord,
  byId: Map<string, WorkspaceTaskRecord>,
): TaskDependencyBlocker[] {
  const deps = Array.isArray(task.dependencies) ? task.dependencies : [];
  const blockers: TaskDependencyBlocker[] = [];
  for (const depId of deps) {
    const cleaned = String(depId || '').trim();
    if (!cleaned) {
      continue;
    }
    const dep = byId.get(cleaned);
    if (!dep) {
      blockers.push({
        taskId: cleaned,
        status: 'missing',
        ownerRole: '',
        summary: 'missing dependency',
      });
      continue;
    }
    const status = String(dep.status || 'open').trim().toLowerCase();
    if (status !== 'completed') {
      blockers.push({
        taskId: cleaned,
        status,
        ownerRole: String(dep.owner_role || '')
          .trim()
          .toLowerCase(),
        summary: String(dep.goal || cleaned).slice(0, 120),
      });
    }
  }
  return blockers;
}

/** Matches control-plane `dependencies_completed` — only `completed` satisfies a dep. */
export function taskDependenciesCompleted(
  task: WorkspaceTaskRecord,
  byId: Map<string, WorkspaceTaskRecord>,
): boolean {
  return taskDependencyBlockers(task, byId).length === 0;
}

export function taskDependencyBlockerMessage(
  task: WorkspaceTaskRecord,
  byId: Map<string, WorkspaceTaskRecord>,
): string {
  const blockers = taskDependencyBlockers(task, byId);
  if (!blockers.length) {
    return '';
  }
  const parts = blockers.slice(0, 3).map((row) => {
    const role = row.ownerRole || 'task';
    return `${role} ${row.taskId} (${row.status})`;
  });
  const suffix = blockers.length > 3 ? ` (+${blockers.length - 3} more)` : '';
  return `blocked by unfinished dependencies: ${parts.join(', ')}${suffix}`;
}

export function humanizeHandoffBlockedReason(
  task: WorkspaceTaskRecord,
  byId: Map<string, WorkspaceTaskRecord>,
): string {
  const blockers = taskDependencyBlockers(task, byId);
  if (!blockers.length) {
    return '';
  }
  const first = blockers[0];
  const roleLabel = first.ownerRole || 'teammate';
  if (first.summary.toLowerCase().startsWith('verification after')) {
    return `Waiting on ${roleLabel} verification before this handoff can start.`;
  }
  return `Waiting on ${roleLabel} (${first.status}) before Start now is available.`;
}
